--- origsrc/ffi-1.8.1/ext/ffi_c/LongDouble.c	2013-06-30 23:08:55.816577800 -0500
+++ src/ffi-1.8.1/ext/ffi_c/LongDouble.c	2013-05-07 03:57:46.000000000 -0500
@@ -41,7 +41,7 @@ rbffi_num2longdouble(VALUE value)
 
     if (RTEST(rb_cBigDecimal) && rb_cBigDecimal != rb_cObject && RTEST(rb_obj_is_kind_of(value, rb_cBigDecimal))) {
         VALUE s = rb_funcall(value, rb_intern("to_s"), 1, rb_str_new2("E"));
-        return strtold(RSTRING_PTR(s), NULL);
+        return (long double) strtod(RSTRING_PTR(s), NULL);
     }
 
     /* Fall through to handling as a float */
--- origsrc/ffi-1.8.1/lib/ffi/library.rb	2013-05-07 03:57:44.462888900 -0500
+++ src/ffi-1.8.1/lib/ffi/library.rb	2013-05-07 04:09:36.353606800 -0500
@@ -38,7 +38,13 @@ module FFI
 
     if lib && File.basename(lib) == lib
       lib = Platform::LIBPREFIX + lib unless lib =~ /^#{Platform::LIBPREFIX}/
-      r = Platform::IS_GNU ? "\\.so($|\\.[1234567890]+)" : "\\.#{Platform::LIBSUFFIX}$"
+      r = if Platform::IS_GNU 
+             "\\.so($|\\.[1234567890]+)"
+          elsif Platform::IS_CYGWIN
+             "\\-[1234567890]+.dll$"
+          else
+             "\\.#{Platform::LIBSUFFIX}$"
+          end
       lib += ".#{Platform::LIBSUFFIX}" unless lib =~ /#{r}/
     end
 
--- origsrc/ffi-1.8.1/lib/ffi/platform/i386-cygwin/types.conf	2013-05-07 03:57:44.477889800 -0500
+++ src/ffi-1.8.1/lib/ffi/platform/i386-cygwin/types.conf	2013-06-30 23:05:50.596057800 -0500
@@ -1,3 +1,40 @@
-rbx.platform.typedef.size_t = uint
-rbx.platform.typedef.ptrdiff_t = int
-rbx.platform.typedef.ssize_t = int
+rbx.platform.typedef.blkcnt_t = long_long
+rbx.platform.typedef.blksize_t = int
+rbx.platform.typedef.clockid_t = long
+rbx.platform.typedef.dev_t = uint
+rbx.platform.typedef.fsblkcnt_t = ulong
+rbx.platform.typedef.fsfilcnt_t = ulong
+rbx.platform.typedef.gid_t = uint
+rbx.platform.typedef.id_t = uint
+rbx.platform.typedef.in_addr_t = uint
+rbx.platform.typedef.in_port_t = ushort
+rbx.platform.typedef.ino_t = ulong_long
+rbx.platform.typedef.int8_t = char
+rbx.platform.typedef.int16_t = short
+rbx.platform.typedef.int32_t = int
+rbx.platform.typedef.int64_t = long_long
+rbx.platform.typedef.key_t = long_long
+rbx.platform.typedef.loff_t = long_long
+rbx.platform.typedef.mode_t = uint
+rbx.platform.typedef.nlink_t = ushort
+rbx.platform.typedef.off_t = long
+rbx.platform.typedef.pid_t = int
+rbx.platform.typedef.pthread_key_t = pointer
+rbx.platform.typedef.pthread_once_t = pointer
+rbx.platform.typedef.pthread_t = pointer
+rbx.platform.typedef.quad_t = long_long
+rbx.platform.typedef.register_t = int
+rbx.platform.typedef.rlim_t = ulong
+rbx.platform.typedef.sa_family_t = ushort
+rbx.platform.typedef.size_t = ulong
+rbx.platform.typedef.socklen_t = int
+rbx.platform.typedef.ssize_t = long
+rbx.platform.typedef.suseconds_t = long
+rbx.platform.typedef.time_t = long
+rbx.platform.typedef.timer_t = ulong
+rbx.platform.typedef.u_int16_t = ushort
+rbx.platform.typedef.u_int32_t = uint
+rbx.platform.typedef.u_int64_t = ulong_long
+rbx.platform.typedef.u_int8_t = uchar
+rbx.platform.typedef.u_quad_t = ulong_long
+rbx.platform.typedef.uid_t = uint
--- origsrc/ffi-1.8.1/lib/ffi/platform/x86_64-cygwin/types.conf	1969-12-31 18:00:00.000000000 -0600
+++ src/ffi-1.8.1/lib/ffi/platform/x86_64-cygwin/types.conf	2013-06-30 23:08:39.617520800 -0500
@@ -0,0 +1,40 @@
+rbx.platform.typedef.blkcnt_t = long
+rbx.platform.typedef.blksize_t = int
+rbx.platform.typedef.clockid_t = long
+rbx.platform.typedef.dev_t = uint
+rbx.platform.typedef.fsblkcnt_t = ulong
+rbx.platform.typedef.fsfilcnt_t = ulong
+rbx.platform.typedef.gid_t = uint
+rbx.platform.typedef.id_t = uint
+rbx.platform.typedef.in_addr_t = uint
+rbx.platform.typedef.in_port_t = ushort
+rbx.platform.typedef.ino_t = ulong
+rbx.platform.typedef.int8_t = char
+rbx.platform.typedef.int16_t = short
+rbx.platform.typedef.int32_t = int
+rbx.platform.typedef.int64_t = long
+rbx.platform.typedef.key_t = long
+rbx.platform.typedef.loff_t = long
+rbx.platform.typedef.mode_t = uint
+rbx.platform.typedef.nlink_t = ushort
+rbx.platform.typedef.off_t = long
+rbx.platform.typedef.pid_t = int
+rbx.platform.typedef.pthread_key_t = pointer
+rbx.platform.typedef.pthread_once_t = pointer
+rbx.platform.typedef.pthread_t = pointer
+rbx.platform.typedef.quad_t = long
+rbx.platform.typedef.register_t = int
+rbx.platform.typedef.rlim_t = ulong
+rbx.platform.typedef.sa_family_t = ushort
+rbx.platform.typedef.size_t = ulong
+rbx.platform.typedef.socklen_t = int
+rbx.platform.typedef.ssize_t = long
+rbx.platform.typedef.suseconds_t = long
+rbx.platform.typedef.time_t = long
+rbx.platform.typedef.timer_t = ulong
+rbx.platform.typedef.u_int16_t = ushort
+rbx.platform.typedef.u_int32_t = uint
+rbx.platform.typedef.u_int64_t = ulong
+rbx.platform.typedef.u_int8_t = uchar
+rbx.platform.typedef.u_quad_t = ulong
+rbx.platform.typedef.uid_t = uint
--- origsrc/ffi-1.8.1/lib/ffi/platform.rb	2013-05-07 03:57:44.683901600 -0500
+++ src/ffi-1.8.1/lib/ffi/platform.rb	2013-05-07 04:10:03.070134900 -0500
@@ -74,6 +74,7 @@ module FFI
     IS_FREEBSD = is_os("freebsd")
     IS_OPENBSD = is_os("openbsd")
     IS_WINDOWS = is_os("windows")
+    IS_CYGWIN = is_os("cygwin")
     IS_BSD = IS_MAC || IS_FREEBSD || IS_OPENBSD
     CONF_DIR = File.join(File.dirname(__FILE__), 'platform', NAME)
     public
